File: android/app/src/main/kotlin/com/example/netscope/network/SignalInfo.kt
package com.example.netscope.network

enum class SignalInfoStatus {
    AVAILABLE,
    PERMISSION_DENIED,
    UNSUPPORTED,
    UNAVAILABLE,
    ERROR,
}

enum class SignalTechnology {
    UNKNOWN,
    LTE,
    NR,
}

data class SignalInfo(
    val status: SignalInfoStatus,
    val technology: SignalTechnology = SignalTechnology.UNKNOWN,
    val rsrp: Int? = null,
    val rsrq: Int? = null,
    val sinr: Double? = null,
    val pci: Int? = null,
    val carrier: String? = null,
    val message: String? = null,
) {
    fun toMap(): Map<String, Any?> {
        return mapOf(
            "status" to status.name.lowercase(),
            "technology" to technology.name.lowercase(),
            "rsrp" to rsrp,
            "rsrq" to rsrq,
            "sinr" to sinr,
            "pci" to pci,
            "carrier" to carrier,
            "message" to message,
        )
    }

    companion object {
        fun available(
            technology: SignalTechnology,
            rsrp: Int?,
            rsrq: Int?,
            sinr: Double?,
            pci: Int?,
            carrier: String? = null,
            message: String? = null,
        ): SignalInfo {
            return SignalInfo(
                status = SignalInfoStatus.AVAILABLE,
                technology = technology,
                rsrp = rsrp,
                rsrq = rsrq,
                sinr = sinr,
                pci = pci,
                carrier = carrier,
                message = message,
            )
        }

        fun permissionDenied(message: String? = null): SignalInfo {
            return SignalInfo(
                status = SignalInfoStatus.PERMISSION_DENIED,
                message = message,
            )
        }

        fun unsupported(message: String? = null): SignalInfo {
            return SignalInfo(
                status = SignalInfoStatus.UNSUPPORTED,
                message = message,
            )
        }

        fun unavailable(message: String? = null): SignalInfo {
            return SignalInfo(
                status = SignalInfoStatus.UNAVAILABLE,
                message = message,
            )
        }

        fun error(message: String? = null): SignalInfo {
            return SignalInfo(
                status = SignalInfoStatus.ERROR,
                message = message,
            )
        }
    }
}

File: lib/models/signal_info.dart
enum SignalInfoStatus {
  available,
  permissionDenied,
  unsupported,
  unavailable,
  error,
}

enum RadioTechnology {
  unknown('Unknown'),
  lte('LTE'),
  nr('5G NR');

  const RadioTechnology(this.displayName);
  final String displayName;
}

extension SignalInfoStatusDisplayName on SignalInfoStatus {
  String get displayName {
    switch (this) {
      case SignalInfoStatus.available:
        return 'available';
      case SignalInfoStatus.permissionDenied:
        return 'permissionDenied';
      case SignalInfoStatus.unsupported:
        return 'unsupported';
      case SignalInfoStatus.unavailable:
        return 'unavailable';
      case SignalInfoStatus.error:
        return 'error';
    }
  }
}

class SignalInfo {
  const SignalInfo({
    required this.status,
    this.technology = RadioTechnology.unknown,
    this.rsrp,
    this.rsrq,
    this.sinr,
    this.pci,
    this.carrier,
    this.message,
  });

  final SignalInfoStatus status;
  final RadioTechnology technology;
  final int? rsrp;
  final int? rsrq;
  final double? sinr;
  final int? pci;
  final String? carrier;
  final String? message;

  bool get hasSignalValues =>
      rsrp != null || rsrq != null || sinr != null || pci != null;

  Map<String, Object?> toMap() {
    return {
      'status': status.displayName,
      'technology': technology.name,
      'rsrp': rsrp,
      'rsrq': rsrq,
      'sinr': sinr,
      'pci': pci,
      'carrier': carrier,
      'message': message,
    };
  }

  factory SignalInfo.fromMap(Map<String, Object?> map) {
    return SignalInfo(
      status: _signalInfoStatusFromName(map['status'] as String?),
      technology: _radioTechnologyFromName(map['technology'] as String?),
      rsrp: (map['rsrp'] as num?)?.toInt(),
      rsrq: (map['rsrq'] as num?)?.toInt(),
      sinr: (map['sinr'] as num?)?.toDouble(),
      pci: (map['pci'] as num?)?.toInt(),
      carrier: map['carrier'] as String?,
      message: map['message'] as String?,
    );
  }
}

SignalInfoStatus _signalInfoStatusFromName(String? name) {
  switch (name) {
    case 'available':
      return SignalInfoStatus.available;
    case 'permissionDenied':
    case 'permission_denied':
      return SignalInfoStatus.permissionDenied;
    case 'unsupported':
      return SignalInfoStatus.unsupported;
    case 'unavailable':
      return SignalInfoStatus.unavailable;
    case 'error':
    default:
      return SignalInfoStatus.error;
  }
}

RadioTechnology _radioTechnologyFromName(String? name) {
  switch (name) {
    case 'lte':
      return RadioTechnology.lte;
    case 'nr':
      return RadioTechnology.nr;
    case 'unknown':
    default:
      return RadioTechnology.unknown;
  }
}
